### PR TITLE
Type helper cleanup, library reorganizations

### PR DESCRIPTION
**Pull request by kevmoo**
Reorganize libraries so there is one clear import for each use case.
Support nesting in `TypeHelper`.
Cleanup the API for `TypeHelper`
Other misc

### CODE REVIEW

**Comment by kevmoo** [comment]
@natebosch it's big, but it's really nice

CC @jakemac53

**Comment by kevmoo** [comment]
@natebosch – rebased on latest

File: lib/src/type_helper.dart [hunk 1]
```diff
@@ -2,25 +2,27 @@
...
-
 import 'package:source_gen/source_gen.dart' show TypeChecker;
 
-abstract class TypeHelper {
-  const TypeHelper();
+/// If [type] is the [Type] or implements the [Type] represented by [checker],
+/// returns the generic arguments to the [checker] [Type] if there are any.
+///
+/// If the [checker] [Type] doesn't have generic arguments, `null` is returned.
+List<DartType> getTypeArguments(DartType type, TypeChecker checker) {
```
Comment (natebosch) [inline lib/src/type_helper.dart]: [nit] could be `typeArguments` or `typeArgumentsOf`... `get` doesn't add value most of the time.

File: lib/src/type_helper.dart [hunk 2]
```diff
@@ -54,68 +60,44 @@ abstract class TypeHelper {
...
-class JsonHelper extends TypeHelper {
-  const JsonHelper();
-
-  //TODO(kevmoo): This should be checking for toJson method, but toJson might be
-  //   gone during generation, so we'll have to check for the annotation, too!
-  // In the mean time, just assume the `canSerialize` logic will work most of
-  //   the time.
-  @override
-  bool canSerialize(DartType type) => canDeserialize(type);
+/// A [TypeChecker] for [String].
```
Comment (natebosch) [inline lib/src/type_helper.dart]: [nit] Is the comment adding new information?

BTW a pattern I've seen Matan use with `TypeChecker` is to name the variable like `$String` which makes the usage look nicer than repeating `TypeChecker`

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,27 @@
 ## 0.2.0
 
+* **BREAKING** Types are now segmented into their own libraries.
+
```
Comment (kevmoo) [inline CHANGELOG.md]: A bit easier to read the source and the parsed markdown.

File: lib/src/json_serializable_generator.dart [hunk 4]
```diff
@@ -12,14 +12,24 @@ import 'package:source_gen/source_gen.dart';
...
 import 'json_serializable.dart';
 import 'type_helper.dart';
+import 'type_helpers/core_helper.dart';
+import 'type_helpers/date_time_helper.dart';
+import 'type_helpers/json_helper.dart';
+import 'type_helpers/list_helper.dart';
+import 'type_helpers/map_helper.dart';
 import 'utils.dart';
 
+final _substitute = "e";
```
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: done

File: lib/src/json_serializable_generator.dart [hunk 5]
```diff
@@ -12,14 +12,24 @@ import 'package:source_gen/source_gen.dart';
...
 // TODO: toJson option to omit null/empty values
 class JsonSerializableGenerator
     extends GeneratorForAnnotation<JsonSerializable> {
+  static const _coreHelpers = const [const IterableHelper(), const MapHelper()];
+
   static const _defaultHelpers = const [
     const JsonHelper(),
-    const DateTimeHelper()
+    const DateTimeHelper(),
+    const CoreHelper()
```
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: Done. Also caught a bug! – CoreHelper should be in _coreHelpers!

File: lib/src/json_serializable_generator.dart [hunk 6]
```diff
@@ -228,196 +238,35 @@ class JsonSerializableGenerator
...
-    // So the only safe types for key are dynamic/Object/String
-    var safeKey = keyArg.isDynamic ||
-        keyArg.isObject ||
-        _coreStringChecker.isExactlyType(keyArg);
-
-    var safeValue = valueType.isDynamic ||
-        valueType.isObject ||
-        _stringBoolNumChecker.isAssignableFromType(valueType);
+  String _serialize(DartType targetType, String expression) => _allHelpers
+      .map((th) => th.serialize(targetType, expression, _serialize))
```
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: `TypeHelper` – would you prefer `e`?

File: lib/src/type_helper.dart (hunk 1 above)
Comment (kevmoo) [inline lib/src/type_helper.dart]: nice!

File: lib/src/json_serializable_generator.dart (hunk 6 above)
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: Maybe just `h` for 'helper' coming from `_allHelpers` -> or change it to `_allTypeHelpers` and keep `th`

File: lib/src/type_helper.dart (hunk 2 above)
Comment (kevmoo) [inline lib/src/type_helper.dart]: Uh...I think this is deleted code...

Comment (kevmoo) [inline lib/src/type_helper.dart]: Nevermind. Found it. Uh...it makes it clear what it's for.

Better a trivial doc comment than no doc comment. This is a public API.

Not sure where you're going with the `$String` thing...

**Pull request merged**